packages/flutter/lib/src/foundation/assertions.dart

```diff
@@ -198,24 +198,56 @@ class FlutterError extends AssertionError {
         debugPrint('The number ${details.exception} was $verb.', wrapWidth: _kWrapWidth);
       } else {
         String errorName;
+        String longMessage;
         if (details.exception is AssertionError) {
           errorName = 'assertion';
+          // Regular _AssertionErrors thrown by assert() put the message last, after
+          // some code snippets. This leads to ugly messages. To avoid this, we move
+          // the assertion message up to before the code snippets, separated by a
+          // newline, if we recognise that format is being used.
+          final String message = details.exception.message;
+          final String fullMessage = details.exception.toString();
+          if (message is String && message != fullMessage) {
+            if (fullMessage.length > message.length) {
+              final int position = fullMessage.lastIndexOf(message);
+              if (position == fullMessage.length - message.length &&
```

Comment:
The AssertionError toString() method doesn't guarantee formatting. Would it be possible to write a test that made this assumption explicit (and failed if it was violated)?